src/gateway/adhoc.cpp

```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
 }
 
 auto
-adhoc_t::resolve(const partition_t& name) const -> std::vector<asio::ip::tcp::endpoint> {
-    remote_map_t::const_iterator lb, ub;
-
-    auto ptr = m_remotes.synchronize();
-
-    if(!ptr->count(name)) {
-        throw std::system_error(error::service_not_available);
-    }
-
-    std::tie(lb, ub) = ptr->equal_range(name);
-
-    std::uniform_int_distribution<int> distribution(0, std::distance(lb, ub) - 1);
-    std::advance(lb, distribution(m_random_generator));
-
-    COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", attribute_list({
-        {"uuid", lb->second.uuid}
-    }));
-
-    return lb->second.endpoints;
+adhoc_t::resolve(const std::string& name) const -> service_description_t {
+    return m_remotes.apply([&](const remote_map_t& remotes) {
+        auto by_service_it = remotes.find(name);
+        if(by_service_it == remotes.end() || by_service_it->second.empty()) {
+            throw std::system_error(error::service_not_available);
+        }
+
+        // try to find local service
+        auto& services_by_uuid = by_service_it->second;
+        auto it = services_by_uuid.find({});
```

Comment:
I can add a constant local_uuid if it helps.
2 maps are easier to use and understand(from my point of view), besides they do not allow duplicate uuid for the same service. Code with multimap with this check looks bulky.